pkg/controllers/coscheduling/podgroup_controller.go

```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+	pg := &schedv1alpha1.PodGroup{}
+	err := r.Get(ctx, req.NamespacedName, pg)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			log.V(5).Info("Pod group has been deleted")
+			return ctrl.Result{}, nil
+		}
+		log.V(3).Error(err, "Unable to retrieve pod group")
+		return ctrl.Result{}, err
+	}
+
+	if pg.Status.Phase == schedv1alpha1.PodGroupFinished ||
+		pg.Status.Phase == schedv1alpha1.PodGroupFailed {
+		return ctrl.Result{}, nil
+	}
+	// If startScheduleTime - createTime > 2days,
+	// do not enqueue again because pod may have been GCed
+	if pg.Status.Scheduled == pg.Spec.MinMember && pg.Status.Running == 0 &&
+		pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour {
+		return ctrl.Result{}, nil
+	}
+
+	pgCopy := pg.DeepCopy()
+	podList := &corev1.PodList{}
+	err = r.List(ctx, podList,
+		client.MatchingLabelsSelector{
+			Selector: labels.Set(map[string]string{
+				schedv1alpha1.PodGroupLabel: pgCopy.Name}).AsSelector(),
+		})
+	if err != nil {
+		log.Error(err, "List pods for group failed")
+		return ctrl.Result{}, err
+	}
+	pods := podList.Items
+
+	switch pgCopy.Status.Phase {
+	case "":
+		pgCopy.Status.Phase = schedv1alpha1.PodGroupPending
+	case schedv1alpha1.PodGroupPending:
+		if len(pods) >= int(pg.Spec.MinMember) {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupPreScheduling
+			fillOccupiedObj(pgCopy, &pods[0])
+		}
+	default:
+		var (
+			running   int32 = 0
+			succeeded int32 = 0
+			failed    int32 = 0
+		)
+		if len(pods) != 0 {
+			for _, pod := range pods {
+				switch pod.Status.Phase {
+				case v1.PodRunning:
+					running++
+				case v1.PodSucceeded:
+					succeeded++
+				case v1.PodFailed:
+					failed++
+				}
+			}
+		}
+		pgCopy.Status.Failed = failed
+		pgCopy.Status.Succeeded = succeeded
+		pgCopy.Status.Running = running
 
-	// TODO(user): your logic here
+		if len(pods) == 0 {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupPending
+			break
+		}
 
-	return ctrl.Result{}, nil
+		if pgCopy.Status.Scheduled >= pgCopy.Spec.MinMember && pgCopy.Status.Phase == schedv1alpha1.PodGroupScheduling {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupScheduled
+		}
+
+		if pgCopy.Status.Succeeded+pgCopy.Status.Running >= pg.Spec.MinMember && pgCopy.Status.Phase == schedv1alpha1.PodGroupScheduled {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupRunning
+		}
+		// Final state of pod group
+		if pgCopy.Status.Failed != 0 && pgCopy.Status.Failed+pgCopy.Status.Running+pgCopy.Status.Succeeded >= pg.Spec.
+			MinMember {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupFailed
+		}
+		if pgCopy.Status.Succeeded >= pg.Spec.MinMember {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupFinished
+		}
+	}
+
+	err = r.patchPodGroup(ctx, pg, pgCopy)
+	return ctrl.Result{}, err
+}
+
+func (r *PodGroupReconciler) patchPodGroup(ctx context.Context, old, new *schedv1alpha1.PodGroup) error {
```

Comment:
make this return `ctrl.Result, error` so you can just `return r.patchPodGroup` in `Reconcile`.

Reply:
actually, I may put the `patch` inside a `defer` earlier, will change this later